src/cluster.py

```diff
@@ -67,16 +67,19 @@ async def _get_nodes(  # pylint: disable=too-many-arguments
     endpoint: str,
     username: str,
     password: str,
+    cacert: str,
     model_name: str,
-    apps_string: Optional[str] = None,
-    units_string: Optional[str] = None,
-    machines_string: Optional[str] = None,
+    apps_string: str = "",
```

Comment:
**nitpick:** 

Why `_string` as suffix?
I think type annotation is better way to describe what this variable means.

Reply:
It's because inside the code:

```python
if apps_string != "":
  apps = parse_comma_separated_strings(apps_string)
  nodes.update(_get_ip_by_apps(model, apps))
```

The type annotation of `_get_ip_by_apps` is `List[str]` and you duplicate definite the variable `apps`.

Ok, if it's the case. Then it's not blocking but you can keep variable name `app_string` and use `Optional[str]` type annootation.